File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/usecase/UseCasesExposeSingleInvoke.kt
package com.perrystreet.woof.konsist.usecase

import com.lemonappdev.konsist.api.ext.list.modifierprovider.withPublicOrDefaultModifier
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class UseCasesExposeSingleInvoke : BehaviorSpec() {
    init {
        Given("A UseCase") {
            val useCases = KonsistUtils.useCases

            Then("Its only public function is operator fun invoke") {
                useCases.assertTrue(message = Message) { useCase ->
                    val publicFunctions = useCase.functions().withPublicOrDefaultModifier()
                    publicFunctions.size == 1 && publicFunctions.single().let { it.name == "invoke" && it.hasOperatorModifier }
                }
            }

            Then("It exposes no public properties") {
                useCases.assertTrue(message = Message) { useCase ->
                    useCase.properties().withPublicOrDefaultModifier().all { it.hasConstModifier }
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "UseCases expose exactly one operator fun invoke() and no public state.",
            why = """
                One class, one use case. A single invoke keeps the responsibility obvious and lets callers
                use the class like a function. Public state would turn it into a service.
            """.trimIndent(),
            howToFix = "Split additional public functions into their own UseCases; move state into a Repository.",
            badExample = "class DogsUseCase { fun load() {}; fun block() {} }",
            goodExample = "class LoadNextDogsPageUseCase { operator fun invoke(): Single<DogsPage> }",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/UseCase/UseCasesExposeSingleInvoke.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class UseCasesExposeSingleInvoke: QuickSpec {
    override class func spec() {
        Given("A UseCase") {
            let useCases = WoofHarmonize.useCases

            Then("Its only public function is callAsFunction") {
                useCases.assertTrue(rule: rule) { useCase in
                    let publicFunctions = useCase.functions.withoutModifier(.private)
                    return publicFunctions.count == 1 && publicFunctions[0].name == "callAsFunction"
                }
            }

            Then("It exposes no public state") {
                useCases.assertTrue(rule: rule) { useCase in
                    useCase.variables.withPublicModifier().allSatisfy { $0.modifiers.contains(.static) && $0.isConstant }
                }
            }
        }
    }

    private static let rule = Rule(
        description: "UseCases expose exactly one callAsFunction() and no public state.",
        rationale: """
            One class, one use case. A single callAsFunction keeps the responsibility obvious and lets callers
            use the class like a function. Public state would turn it into a service.
            """,
        fixHint: "Split additional public functions into their own UseCases; move state into a Repository.",
        badExample: "final class DogsUseCase { func load() {}; func block() {} }",
        goodExample: "final class LoadNextDogsPageUseCase { func callAsFunction() -> AnyPublisher<DogsPage, DogsError> }"
    )
}
